Take \tikzpicture out of karnaugh-map environment (fixes #21)

myzhang1029 opened the pull request:
I stumbled upon #21 myself and found [this answer](https://tex.stackexchange.com/questions/352260/tikzexternalize-conflicts-with-begincircuitikz) that claims that `tikzexternalize` scans for `end{tikzpicture}`.

This patch allows the user to place their own `tikzpicture` environment around the `karnaugh-map-inner` environment, which makes the package compatible with tikz externalization.

If the code looks reasonable, then I am also happy to add the documentation.

In this test, the results with or without externalizing (by toggling `\def\testingexternalize` are visually identical.
```
\documentclass{article}
\usepackage{karnaugh-map}
\usetikzlibrary{external}
\def\testingexternalize{0}

\if\testingexternalize1
    \tikzexternalize[prefix=kmmwe-]
\fi

\begin{document}
\if\testingexternalize1
    \begin{tikzpicture}
    \begin{karnaugh-map-inner}[4][4][1][$D$][$C$][$B$][$A$]
        \minterms{0,2,3,5,7,8,9,10,11,13,15}
        \implicant{5}{15}
        \implicantcorner{}
        \implicantedge{3}{2}{11}{10}
    \end{karnaugh-map-inner}
    \end{tikzpicture}
\else
    \begin{karnaugh-map}[4][4][1][$D$][$C$][$B$][$A$]
        \minterms{0,2,3,5,7,8,9,10,11,13,15}
        \implicant{5}{15}
        \implicantcorner{}
        \implicantedge{3}{2}{11}{10}
    \end{karnaugh-map}
\fi
\end{document}
```

myzhang1029 committed:
take \tikzpicture out of karnaugh-map environment

This allows the user to place their own tikzpicture environment around
the karnaugh-map-inner environment, which makes the package compatible
with tikz externalization.

myzhang1029 committed:
Remove the now-unneeded group and add test

Signed-off-by: Zhang Maiyun <me@maiyun.me>

2pi commented:
Thank you for the PR! I've looked briefly at your patch, and I'm not entirely happy with introducing yet another environment(`karnaugh-map-inner`).

Trying to follow the design-pattern of the package, this would probably be most suitable if it were introduced as an _option_; see [1.3 Options](https://mirrors.ctan.org/graphics/pgf/contrib/karnaugh-map/karnaugh-map.pdf#page=8) in the documentation. For example as `omittikzpicture` or something better. It would, if I'm not mistaken, also make it easier to switch back and fourth.

Any thoughts?

myzhang1029 commented:
Thanks!
Yes, I definitely agree that an entire new environment is a bit too messy. Changing this to an option should be doable.

Maiyun Zhang (he/him)

> On 7 Mar 2025, at 10:49, 2pi ***@***.***> wrote:
> 
> 
> 2pi
>  left a comment 
> (2pi/karnaugh-map#24)
> Thank you for the PR! I've looked briefly at your patch, and I'm not entirely happy with introducing yet another environment(karnaugh-map-inner).
> 
> Trying to follow the design-pattern of the package, this would probably be most suitable if it were introduced as an option; see 1.3 Options in the documentation. For example as omittikzpicture or something better. It would, if I'm not mistaken, also make it easier to switch back and fourth.
> 
> Any thoughts?
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.
> 
>  <https://mirrors.ctan.org/graphics/pgf/contrib/karnaugh-map/karnaugh-map.pdf#page=8> <https://github.com/2pi/karnaugh-map/pull/24#issuecomment-2707176729> <https://github.com/notifications/unsubscribe-auth/AFSX55FD5S4O2WACRE3KYRL2THS3ZAVCNFSM6AAAAABTD5JTQWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMBXGE3TMNZSHE>
> 
> 2pi
>  left a comment 
> (2pi/karnaugh-map#24)
>  <https://github.com/2pi/karnaugh-map/pull/24#issuecomment-2707176729>
> Thank you for the PR! I've looked briefly at your patch, and I'm not entirely happy with introducing yet another environment(karnaugh-map-inner).
> 
> Trying to follow the design-pattern of the package, this would probably be most suitable if it were introduced as an option; see 1.3 Options <https://mirrors.ctan.org/graphics/pgf/contrib/karnaugh-map/karnaugh-map.pdf#page=8> in the documentation. For example as omittikzpicture or something better. It would, if I'm not mistaken, also make it easier to switch back and fourth.
> 
> Any thoughts?
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/2pi/karnaugh-map/pull/24#issuecomment-2707176729>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AFSX55FD5S4O2WACRE3KYRL2THS3ZAVCNFSM6AAAAABTD5JTQWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMBXGE3TMNZSHE>.
> You are receiving this because you authored the thread.
>

myzhang1029 committed:
Convert omittikzpicture to an option

Signed-off-by: Zhang Maiyun <me@maiyun.me>

myzhang1029 committed:
Add test for the global option

Signed-off-by: Zhang Maiyun <me@maiyun.me>

2pi commented on karnaugh-map.dtx:
```diff
@@ -244,10 +244,12 @@
 %% environment options
 \define@key{karnaugh-map}{label}[middle]{\renewcommand{\@karnaughmap@option@label}{#1}}
 \define@key{karnaugh-map}{implicantcolors}[]{\renewcommand{\@karnaughmap@option@implicantcolors}{#1}}
+\define@key{karnaugh-map}{omittikzpicture}[false]{\@karnaughmap@option@omittikzpicturetrue}
```
please move this line down one line, just to have the same order of options as on line 246-247

2pi commented:
Before continuing, two things. How would you like to be attributed? Would this be okay(hyperlink only when appropriate, see earlier examples)? 

> [Maiyun Zhang](https://github.com/myzhang1029)

Regarding licensing. Similarly to https://github.com/2pi/karnaugh-map/pull/7#issuecomment-277514055 I just want to check that you are aware and okay with the licensing([CC BY-SA](https://creativecommons.org/licenses/by-sa/3.0/)).

myzhang1029 committed:
Reorder options

myzhang1029 committed:
Add documentation for omittikzpicture

myzhang1029 commented:
Thank you. I am fine with the attribution and license.

I have fixed the options and added some draft for the documentation. Does this look good? This option is not very visually visible, so I included sort of a not-so-useful example of a rotated karnaugh map. Please advise if you have better ideas.

myzhang1029 commented:
Also, in the original code, I believe there was extra space after the `\end{tikzpicture}` (should have been a `%`?). Now that we made the `\end{tikzpicture}` optional. For backward compatibility, I kept this extra space if `omittikzpicture` is false. However, this spacing is missing when `omittikzpicture` is specified.

Do you think this is an issue?

Please check this comparison:
[karnaugh-221-label-omit-tikzpicture.pdf](https://github.com/user-attachments/files/19196003/karnaugh-221-label-omit-tikzpicture.pdf)
[karnaugh-221-label.pdf](https://github.com/user-attachments/files/19196007/karnaugh-221-label.pdf)

2pi commented:
> Also, in the original code, I believe there was extra space after the `\end{tikzpicture}` (should have been a `%`?). Now that we made the `\end{tikzpicture}` optional. For backward compatibility, I kept this extra space if `omittikzpicture` is false. However, this spacing is missing when `omittikzpicture` is specified.
> 
> Do you think this is an issue?
> 
> Please check this comparison: [karnaugh-221-label-omit-tikzpicture.pdf](https://github.com/user-attachments/files/19196003/karnaugh-221-label-omit-tikzpicture.pdf) [karnaugh-221-label.pdf](https://github.com/user-attachments/files/19196007/karnaugh-221-label.pdf)

I agree, better to keep it like this for backwards compatibility. Might be something to look at for a future major version. May I ask how you made the examples?

2pi commented on karnaugh-map.dtx:
```diff
@@ -564,14 +576,14 @@
     }%
   \fi
-  \begin{tikzpicture}[
+  \tikzset{
```
I've been thinking a bit about scoping of the options. Using `\tikzset` exposes the options to, what I understand, the whole `tikzpicture`. I found the environment `scope` which can be used inside a `tikzpicture` to scope things. I'm thinking that it might be better to use it...

    \begin{scope}[<options>]
    \end{scope}

We should be able to use it both with and without the `tikzpicture`...

What do you think?

myzhang1029 replied:
This makes sense. Looks like without `scope`, we would turn the entire outer tikzpicture into karnaugh-map, so it would indeed be useful.

2pi commented:
> Thank you. I am fine with the attribution and license.

Great!

> I have fixed the options and added some draft for the documentation. Does this look good? This option is not very visually visible, so I included sort of a not-so-useful example of a rotated karnaugh map. Please advise if you have better ideas.

I'm thinking that an simple example would be better. Like a plain `karnaugh-map` environment without anything inside. Probably even just showing the code instead of the rendering for this one as this option is technical and not visual.

2pi commented on karnaugh-map.dtx:
```diff
@@ -1462,6 +1474,17 @@
 %   \end{karnaugh-map}
 % \end{figure}
 %
+% \texttt{omittikzpicture}: By default, the |karnaugh-map| environment creates its own |tikzpicture|. This option disables this behavior, and it would be expected that the user wraps the |karnaugh-map| environment in a |tikzpicture| environment. This is useful to make the |tikzexternal| package work with |karnaugh-map|, or to integrate |karnaugh-map| into a larger |tikzpicture|.
```
> By default, the |karnaugh-map| environment creates its own |tikzpicture| **environment**. This option disables this behavior, and it would be expected that the user wraps the |karnaugh-map| environment in a |tikzpicture| environment. This **option makes it possible to use the** |tikzexternal| **package with** |karnaugh-map|, or to integrate |karnaugh-map| into a larger |tikzpicture|.

Some minor input...

2pi commented:
Regarding testing, I think it would be best to do something similar to `karnaugh-option-implicantcolors-default.tex`  `karnaugh-option-implicantcolors.tex` while alternating the usage of the option. Both files with the *exact* same rendering(adding extra spacing as needed, I'm referring to what the above mentioned issue) but the first one with mix of using the option and not(not using the option, using the option, not using the option). While the second one uses the global option but the rest remains the same. What do you think?

myzhang1029 commented:
> I agree, better to keep it like this for backwards compatibility. Might be something to look at for a future major version. May I ask how you made the examples?

They were produced with the [`lua-visual-debug`](https://www.ctan.org/pkg/lua-visual-debug) package.

myzhang1029 committed:
Update docs per suggestion

Signed-off-by: Zhang Maiyun <me@maiyun.me>

myzhang1029 committed:
Use scoping

Signed-off-by: Zhang Maiyun <me@maiyun.me>

myzhang1029 committed:
Move scope to main environment

Strangely end{scope} does not produce the same extra space

Signed-off-by: Zhang Maiyun <me@maiyun.me>

myzhang1029 committed:
Code-only example for omittikzpicture

Signed-off-by: Zhang Maiyun <me@maiyun.me>

myzhang1029 committed:
Implement option testing (hack)

Signed-off-by: Zhang Maiyun <me@maiyun.me>

myzhang1029 committed:
Swap misnamed test files

Signed-off-by: Zhang Maiyun <me@maiyun.me>

myzhang1029 committed:
Allow local overrides to global bool options

Signed-off-by: Zhang Maiyun <me@maiyun.me>

myzhang1029 commented:
8ba81d6 might be a little intrusive: it is to allow local environment options to override global ones (i.e. if we had `\usepackage[omittikzpicture, implicantcorners]{karnaugh-map}`, we can now specify `(omittikzpicture=false, implicantcorners=false)` to disable them for one specific environment.

Apparently it wasn't allowed in the previous versions. Please advise if I am doing this incorrectly.

2pi commented:
> [8ba81d6](https://github.com/2pi/karnaugh-map/commit/8ba81d653bde3ba4c4804041f74e49cd07f1bf88) might be a little intrusive: it is to allow local environment options to override global ones (i.e. if we had `\usepackage[omittikzpicture, implicantcorners]{karnaugh-map}`, we can now specify `(omittikzpicture=false, implicantcorners=false)` to disable them for one specific environment.
> 
> Apparently it wasn't allowed in the previous versions. Please advise if I am doing this incorrectly.

Not sure... could be useful but also makes it a bit more complex. Regardless, better if we discuss this in a separate pull-request.

2pi commented:
I've done a bit of changes to everything but the tests in [pr24](https://github.com/2pi/karnaugh-map/tree/pr24) what are your opinions on this?

myzhang1029 commented:
> Not sure... could be useful but also makes it a bit more complex. Regardless, better if we discuss this in a separate pull-request.

I added this to implement this test:
> While the second one uses the global option but the rest remains the same.

Without it, a global `[omittikzpicture]` would permanently enable this option for the entire document. We would have to fiddle with `\@karnaughmap@options@omittikzpicture` if we wanna implement the said test. What do you think? I am fine with not including this part too.

I hope this explanation makes sense though. My point is, it would be impossible to test a "mix of using the option and not" if the global option cannot be locally overridden.

myzhang1029 commented:
Your version in `pr24` seems to better match the original structure of the environment. Looks good.

2pi commented:
Sorry for the delay! Since there are a lot of back and forth in the commits for this PR. Would you mind squashing together something more sensible without the tests of `pr24`? Feel free to include my [last commit](https://github.com/2pi/karnaugh-map/commit/494c41ae6ba2be7b3165c789061a121808a0053b) in the branch too.

Then I will  add some reasonable tests to so that we don't have to go back and fourth and delay it even more. Does it sounds good to you?
